test_communication/test/test_action_client_server.py.in

```diff
@@ -23,41 +23,44 @@ def generate_test_description(action_type):
     launch_description = LaunchDescription()
 
     action_client_cmd = ['@TEST_ACTION_CLIENT_EXECUTABLE@', action_type, namespace]
-    action_server_cmd = ['@TEST_ACTION_SERVER_EXECUTABLE@', action_type, namespace]
-
-    action_server_env = dict(os.environ)
     action_client_env = dict(os.environ)
-
-    if '@TEST_ACTION_SERVER_RCL@' == 'rclpy':
-        action_server_cmd.insert(0, sys.executable)
-        action_server_env['PYTHONUNBUFFERED'] = '1'
     if '@TEST_ACTION_CLIENT_RCL@' == 'rclpy':
         action_client_cmd.insert(0, sys.executable)
         action_client_env['PYTHONUNBUFFERED'] = '1'
-
-    action_server_env['RCL_ASSERT_RMW_ID_MATCHES'] = '@ACTION_SERVER_RMW@'
-    action_server_env['RMW_IMPLEMENTATION'] = '@ACTION_SERVER_RMW@'
-
-    launch_description.add_action(ExecuteProcess(
-        cmd=action_server_cmd,
-        name='test_action_server',
-        env=action_server_env,
-    ))
-
     action_client_env['RCL_ASSERT_RMW_ID_MATCHES'] = '@ACTION_CLIENT_RMW@'
     action_client_env['RMW_IMPLEMENTATION'] = '@ACTION_CLIENT_RMW@'
-
     action_client_process = ExecuteProcess(
         cmd=action_client_cmd,
         name='test_action_client',
         env=action_client_env,
     )
-    launch_description.add_action(action_client_process)
 
-    launch_description.add_action(
-        launch_testing.actions.ReadyToTest()
+    action_server_cmd = ['@TEST_ACTION_SERVER_EXECUTABLE@', action_type, namespace]
+    action_server_env = dict(os.environ)
+    if '@TEST_ACTION_SERVER_RCL@' == 'rclpy':
+        action_server_cmd.insert(0, sys.executable)
+        action_server_env['PYTHONUNBUFFERED'] = '1'
+    action_server_env['RCL_ASSERT_RMW_ID_MATCHES'] = '@ACTION_SERVER_RMW@'
+    action_server_env['RMW_IMPLEMENTATION'] = '@ACTION_SERVER_RMW@'
+
+    action_server_process = ExecuteProcess(
+        cmd=action_server_cmd,
+        name='test_action_server',
+        env=action_server_env,
     )
-    return launch_description, locals()
+
+    return LaunchDescription([
+        # Always stop daemon to isolate tests.
+        ExecuteProcess(
+            cmd=['ros2', 'daemon', 'stop'],
+            name='daemon-stop',
```

Comment:
I wonder why we have a daemon running in the background in the first place.
This looks fine as a workaround, but all previous process that created a daemon should not be "leaking it".

Reply:
Maybe we can have something running in CI that detects when a test "leaks" a ros 2 daemon, but it doesn't sound like something trivial to implement.

Reply:
> I wonder why we have a daemon running in the background in the first place.
> This looks fine as a workaround, but all previous process that created a daemon should not be "leaking it".

In other places in the codebase, we are careful to kill the daemon before running the tests.  For instance: https://github.com/ros2/ros2cli/blob/f6c4a5b05d172f8156645b8fcf8bd2dfa23e7bb2/ros2node/test/test_cli_duplicate_node_names.py#L55 .  From that perspective, it is good for each test to be "defensive", and make sure the environment is the way it should be before starting.

But we are not similarly careful about tests cleaning up after themselves.  In that same example in ros2node, we kill the daemon on startup, but then we do not kill it at the end.  Maybe we should enhance all of the tests do do that, but it is honestly not clear to me.  Partly the problem is that starting up the daemon can be a side-effect of doing other things (like running a `ros2` cli command), so it is not always clear which tests need to kill it at the end.

Reply:
> but then we do not kill it at the end. Maybe we should enhance all of the tests do do that,

:+1: I think that killing the daemon at the end of those tests would be a good idea.

> Partly the problem is that starting up the daemon can be a side-effect of doing other things (like running a ros2 cli command), so it is not always clear which tests need to kill it at the end.

Yes, having a defensive `ros2 daemon stop` before testing doesn't look bad, but it's also not clear which tests can be affected by the presence of a "leaked" daemon and which ones not.
So it sounds like we would actually like to run "ros2 daemon stop" before running each test.